core/src/test/java/org/infinispan/remoting/transport/ControlledTransport.java

```diff
@@ -21,63 +25,81 @@
  * @since 7.0
  */
 public class ControlledTransport extends AbstractDelegatingTransport {
+   private static final Log log = LogFactory.getLog(ControlledTransport.class);
+   private static final Predicate<ReplicableCommand> NEVER = cmd -> false;
 
    private final ReclosableLatch replicationLatch = new ReclosableLatch(true);
    private final ReclosableLatch blockingLatch = new ReclosableLatch(true);
-   private volatile Set<Class> blockBeforeFilter = Collections.emptySet();
-   private volatile Set<Class> blockAfterFilter = Collections.emptySet();
-   private volatile Set<Class> failFilter = Collections.emptySet();
+   private volatile Predicate<ReplicableCommand> blockBeforeFilter = NEVER;
+   private volatile Predicate<ReplicableCommand> blockAfterFilter = NEVER;
+   private volatile Predicate<ReplicableCommand> failFilter = NEVER;
 
-   public ControlledTransport(Transport realOne) {
+   private ControlledTransport(Transport realOne) {
       super(realOne);
    }
 
+   public static ControlledTransport replace(Cache<?, ?> cache) {
+      return wrapGlobalComponent(cache.getCacheManager(), Transport.class, ControlledTransport::new, true);
+   }
+
    @Override
    public void start() {
       //skip start it again.
    }
 
    public void failFor(Class... filter) {
-      this.failFilter = new HashSet<>(Arrays.asList(filter));
+      failFor(classListToFilter(filter));
+   }
+
+   private void failFor(Predicate<ReplicableCommand> filter) {
+      this.failFilter = filter;
       blockingLatch.open();
    }
 
    public void stopFailing() {
-      this.failFilter = Collections.emptySet();
+      this.failFilter = NEVER;
       blockingLatch.open();
    }
 
    public void blockBefore(Class... filter) {
-      this.blockBeforeFilter = new HashSet<>(Arrays.asList(filter));
+      blockBefore(classListToFilter(filter));
+   }
+
+   public void blockBefore(Predicate<ReplicableCommand> filter) {
+      this.blockBeforeFilter = filter;
       replicationLatch.close();
       blockingLatch.close();
    }
 
    public void blockAfter(Class... filter) {
-      this.blockAfterFilter = new HashSet<>(Arrays.asList(filter));
+      blockAfter(classListToFilter(filter));
+   }
+
+   public void blockAfter(Predicate<ReplicableCommand> filter) {
```

Comment:
Without any further info on what caused the latest failures, I'd say it's this... the more complex the filter gets, the more likely it is to exclude the wrong command and crash the whole Rube Goldberg machine.

True, it's painful to enumerate all the commands that state transfer now needs like I'm trying to do with `MockTransport`/`MockLocalTopologyManager`, but IMO those can easily the sequence of commands can be extracted into reusable methods, and where the "business logic" is too complex for the test to predict the commands, then perhaps the logic should be simplified. I'm integrating the PR, but I trust you'll remember what you did to debug the test by the next time it fails :)